### PR TITLE
New quirk for sh1107

### PR DESCRIPTION
**Pull request by mdroberts1243**
This change introduces a new quirk to support the FeatherWing SH1107 128x64 OLED display.  A parameter, "column_and_page_addressing" when set to true in the driver will cause the displayio_display_core_set_region_to_update function to use the column lower, column higher and page set commands on the sh1107.

I wrote a new driver to test with: https://github.com/mdroberts1243/mdroberts1243_CircuitPython_DisplayIO_SH1107

Right now python is throwing an error that I'm sending too many keyword arguments to the super().__init__ 

> Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
Traceback (most recent call last):
  File "code.py", line 32, in <module>
  File "mdroberts1243_displayio_sh1107.py", line 84, in __init__
TypeError: extra keyword arguments given
Press any key to enter the REPL. Use CTRL-D to reload.

### CODE REVIEW

File: shared-module/displayio/display_core.c [hunk 1]
```diff
@@ -283,6 +294,14 @@ void displayio_display_core_set_region_to_update(displayio_display_core_t* self,
...
     }
+    // Quirk for SH1107 "column_and_page_addressing"
+    //      Page address command = 0xB0
+    if (column_and_page_addressing) {
+        data[0] = 0xB0 | (y1 & 0x07);
+        data_length = 1;
+    }
+    self->send(self->bus, data_type, chip_select, data, data_length);
+
     self->send(self->bus, data_type, chip_select, data, data_length);
```
Comment (jepler) [inline shared-module/displayio/display_core.c:305]: I haven't read the datasheet or anything, but this repeated send of the same data looks fishy.  Is it intended?

File: .gitignore [hunk 2]
```diff
@@ -80,6 +80,7 @@ TAGS
 *.mo
 
 .vscode
+.idea
```
Comment (jepler) [inline .gitignore:83]: You can ignore the files that your editor environment creates _in all git repositories_ by creating a "global gitignore file", so that you don't have to add this to every project you collaborate on.  https://docs.github.com/en/github/using-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer

**Review by jepler — COMMENTED** [review]
This seems fine if it is working, though from one of your comments it sounds like it's not working yet.  I'd also want to find out from @tannewt if the "column_and_page_addressing" flag should go in the "core" object or stay where it's been placed.

**Comment by mdroberts1243** [comment]
Thanks Jeff. To clarify, Scott asked me to post a draft PR so he could review and help me get past this problem with the bindings. I don't know what I missed in trying to pass the new parameter from the driver back to the display core.   -mark.

File: shared-module/displayio/display_core.c (hunk 1 above)
Comment (mdroberts1243) [reply shared-module/displayio/display_core.c:305]: There's only supposed to be one send.  I'll check my code again.

File: .gitignore (hunk 2 above)
Comment (mdroberts1243) [reply .gitignore:83]: Thx!

**Comment by tannewt** [comment]
@mdroberts1243 I don't see anything wrong with your code here or in your test library. That error can happen if the name of a kwarg doesn't match one in the parse table. Maybe a typo? Does it go away if you comment out the new kwarg? If so, maybe you don't have your code loaded on the device. You can do `mp_printf(&mp_plat_print, "hello");` in your C code to print out to serial and confirm it is run.

**Comment by mdroberts1243** [comment]
### SH1107 DisplayIO Support Update
I thought I would use this draft PR to keep posting progress on trying to get SH1107 supported under DisplayIO.

**TL;DR:**
I'm stuck because DisplayIO seems to expect the display controller to auto-increment past page (page mode writes) or column (vertical mode writes) boundaries. SH1107 requires you to set the page or column address at the end of each boundary or it will just wrap and start updating the same page or column of memory. I'm not sure this can be worked around and I'm thinking that SH1107 OLED displays should be some kind of FrameBufferIO type display, buffered in memory and then refreshed all at once periodically.

**The Good**
- I've had great support so far in attempting this project! Kudos to both Scott and Jeff.  Especially appreciated Scott's live-stream deep dive into the way DisplayIO is implemented which helped to kick this off.
- I have learned a lot about CircuitPython, the building process, using GItHub, DisplayIO and SH1107 (this started with me placing a Covid-support order on Adafruit to get some stuff to play CircuitPython with ;-) )
- What I (we) set out to do works perfectly.  Setting the quirk (with suitable downgrade in name) to True allowed me to implement conditional code in displayio_display_core_set_region_to_update() that would set upper and lower column addresses and the page address appropriately to the beginning of the subrectangle that was to be painted. I've been able to see bitmaps and text painted onto the OLED using DisplayIO, but only in the swath of a single page (8 pixels deep across the 'width' of 64 columns on the FeatherWing OLED).
- We found a bug in the MP_ARGS Parsing!
And finally:
- I had an excuse to buy an oscilloscope that decodes I2C (didn't need it or use it, but "No job is worth starting unless you can buy a new tool.")

**The Bad**
- DisplayIO seems to expect a display controller that would be able to write any subrectangle of the display's size and just spits out the data continuously once that has been set up.  There doesn't seem to be a way to impose boundaries (like only write one column's worth or one page worth)
- SH1107 only supports two write modes... page and vertical (column).  It will auto-increment across a page (8 pixels deep) for the width of the display (64 columns in the case of the FeatherWing) OR along a column of pages (16 pages, 8 pixels deep for 128 pixels total).  When you reach the boundary the page or column will stay the same and the writes will wrap to the beginning of the page or column and overwrite what you've just done.  You are supposed to set the page or column, depending on mode, each time you reach the boundary.

**The Ugly**
- You could always use the adafruit_framebuf driver I put in place (when the fix to the text() bug propagates through the system). How do I contribute that, BTW?
- It seems to me this kind of display might be best served by a DisplayIO FrameBufferIO driver.  I saw some docs, but it wasn't clear to me whether there are displays using this and how to get started creating one.  I would imagine (dangerous to ASSume) that you could do all the DisplayIO functions on a memory buffer and then automagically paint the screen periodically to update the display.

**Where to go from here?**
- Looking for ideas.
- Am I missing something about DisplayIO? Is it possible for it to honour page/column boundaries and only send one page or one column of data each time?
- Are there FrameBufferIO drivers like I'm imagining?  Can I have a worked example?

Thanks for making it this far!

**Comment by tannewt** [comment]
I think you can add a second quirk (or just call this the sh quirk and have it do two things.) The second thing you could do is change the subrectangle that is updated here: https://github.com/adafruit/circuitpython/blob/main/shared-module/displayio/Display.c#L277 Basically change each subrectangle to be the smallest area you can update.

**Comment by mdroberts1243** [comment]
> I think you can add a second quirk (or just call this the sh quirk and have it do two things.) The second thing you could do is change the subrectangle that is updated here: https://github.com/adafruit/circuitpython/blob/main/shared-module/displayio/Display.c#L277 Basically change each subrectangle to be the smallest area you can update.

Thanks.  I might try switching to 'vertical' mode as well, then the display will have a horizontal orientation naturally without the user having to set it up.  Console output will work better too in that case.

Bandwidth is somewhat limited next few days, so progress might be slow...

**Comment by mdroberts1243** [comment]
I have done a clean up and committed changes  "Mostly-working-version with comments".

I have also updated the driver and the example code: https://github.com/mdroberts1243/mdroberts1243_CircuitPython_DisplayIO_SH1107 (don't know how to contribute that!)

Currently everything looks like it is working except label text (bitmap_label will work fine).  An easy way to see the problem with text is to run the example code and then enter the REPL.  You will see that the terminalio<?> output works fine but when you enter code into the REPL the text you enter will wrap back to the beginning of the line at about the 1/2 way point on the screen.

See the circuitpython discord for a picture.

**Comment by mdroberts1243** [comment]
I fixed the masking error I had in setting the page on the controller and now all the basic operations work. You can use it with the REPL even.

How do I contribute the driver and example code linked above? Not sure how to create a "Pull Request" for something new.

Let me know if you need anything else.

-mark.

**Comment by tannewt** [comment]
> I fixed the masking error I had in setting the page on the controller and now all the basic operations work. You can use it with the REPL even.
> 
> How do I contribute the driver and example code linked above? Not sure how to create a "Pull Request" for something new.
> 
> Let me know if you need anything else.
> 
> -mark.

For the driver we can either fork your copy into the adafruit org or you can transfer it. We'll maintain it since we sell a product with the display. Would you like us to fork your repo?

**Comment by tannewt** [comment]
Also, congrats! You found all of the boards with displays initialized in board_init. You'll need to add extra `, false` to all of those calls to make the compiler happy.

File: shared-module/displayio/Display.c [hunk 3]
```diff
@@ -240,11 +243,23 @@ STATIC bool _refresh_area(displayio_display_obj_t* self, const displayio_area_t*
...
+    // for SH1107 and other boundary constrained controllers
+    //      write one single row at a time
+    if (self->SH1107_addressing) {
+        subrectangles = rows_per_buffer;  // vertical (column mode) write each separately (height times)
+    }
+
+    // Skip this recalculation of subrectangles if SH1107 (each subrectangle is a single row)
+    // [mdroberts1243] I am worried/confused about the pixels_in_byte_share_row calculation though
+    //       since it makes sense to be on a byte boundary (actually a page boundary too)
+    //       seems to work as is though.
```
Comment (tannewt) [inline shared-module/displayio/Display.c:260]: I think it's fine to rely on your testing. pixels_in_byte_share_row should be what you want.

File: shared-module/displayio/Display.c [hunk 4]
```diff
@@ -240,11 +243,23 @@ STATIC bool _refresh_area(displayio_display_obj_t* self, const displayio_area_t*
...
+    if ((displayio_area_size(&clipped) > buffer_size * pixels_per_word)
+                    && (!self->SH1107_addressing)) {
```
Comment (tannewt) [inline shared-module/displayio/Display.c:262]: I think it may be clearer to add an `if` before this one where SH1107_addressing is true. This condition then goes back to normal as an `else if` on the quirk. Then, set both `subrectangles`, like you do above, and `rows_per_buffer` to 1 so you don't need to set `remaining_rows` to 1 each loop below. This approach also matches the `if` body here which sets both of those variables. I think the logic in the loop will just work as it originally was then. The only thing you'll need to add there is passing the quirk status into setting the region.

File: shared-module/displayio/display_core.c [hunk 5]
```diff
@@ -41,7 +41,7 @@
 #include <string.h>
 
 #define DISPLAYIO_CORE_DEBUG(...) (void)0
-// #define DISPLAYIO_CORE_DEBUG(...) mp_printf(&mp_plat_print __VA_OPT__(,) __VA_ARGS__)
+//#define DISPLAYIO_CORE_DEBUG(...) mp_printf(&mp_plat_print __VA_OPT__(,) __VA_ARGS__)
```
Comment (tannewt) [inline shared-module/displayio/display_core.c:44]: Change back.

File: shared-module/displayio/display_core.c [hunk 6]
```diff
@@ -301,7 +322,7 @@ void displayio_display_core_start_refresh(displayio_display_core_t* self) {
 
 void displayio_display_core_finish_refresh(displayio_display_core_t* self) {
     if (self->current_group != NULL) {
-        DISPLAYIO_CORE_DEBUG("displayiocore group_finish_refresh\n");
+//       DISPLAYIO_CORE_DEBUG("displayiocore group_finish_refresh\n");
```
Comment (tannewt) [inline shared-module/displayio/display_core.c:325]: Change back.

**Review by tannewt — CHANGES_REQUESTED** [review]
Really nice work @mdroberts1243, thank you! Just a couple suggestions to make the code clearer. You will also need to fix up the display construct calls for boards with built in displays. Thanks!

**Comment by mdroberts1243** [comment]
> > I fixed the masking error I had in setting the page on the controller and now all the basic operations work. You can use it with the REPL even.
> > How do I contribute the driver and example code linked above? Not sure how to create a "Pull Request" for something new.
> > Let me know if you need anything else.
> > -mark.
> 
> For the driver we can either fork your copy into the adafruit org or you can transfer it. We'll maintain it since we sell a product with the display. Would you like us to fork your repo?

Yes, please fork and maintain! I'm will to transfer as well, but don't know how that is accomplished.

**Comment by mdroberts1243** [comment]
> Also, congrats! You found all of the boards with displays initialized in board_init. You'll need to add extra `, false` to all of those calls to make the compiler happy.

O.k. That was a bit tedious!

Also, there is an issue in build-arm (electronut_labs_blip) where it complains about too few arguments on the set region to update... I suppose they've made their own copy of that for some reason?  I probably need to change that too...

File: shared-module/displayio/Display.c (hunk 3 above)
Comment (mdroberts1243) [reply shared-module/displayio/Display.c:260]: O.k. I might leave the comment in place though.  It seems the byte boundary is taken care of by the pixels_in_byte_share_row setting elsewhere, but the fact that this section claims to align on byte boundaries bothers me, because I'm skipping it!

File: shared-module/displayio/display_core.c (hunk 5 above)
Comment (mdroberts1243) [reply shared-module/displayio/display_core.c:44]: Added the space back

File: shared-module/displayio/display_core.c (hunk 6 above)
Comment (mdroberts1243) [reply shared-module/displayio/display_core.c:325]: Done

File: shared-module/displayio/Display.c (hunk 4 above)
Comment (mdroberts1243) [reply shared-module/displayio/Display.c:262]: I've coded this change but will take some time to test  a bit!

Not sure what you mean by "The only thing you'll need to add there is passing the quirk status into setting the region."  The quirk is already passed into setting the region.

**Comment by mdroberts1243** [comment]
I have made all the requested changes and updates.  But a few boards are still failing the checks but due to messages related to the number of arguments passed to displayio_display_core_set_region_to_update. The errors seem to be the same. Boards include: electronut_labs_blip, pewpew_m4, pybadge, pybadge_airlift, pygamer, pygamer_advance, pyportal, pyportal_titano, ugame10.  Many of these didn't show this problem until after the original common_hal function changes were made.

@tannewt I need help understanding why these errors are being thrown and what is causing them... as far as I know the files they are related to are common to all the boards!  Every board should see the same number of arguments into displayio_display_core_set_region_to_update

Also, in testing the new logic I decided to move a label around the screen by a pixel or two in x and y and see if there were any issues with strange clipping etc.  This heavily loaded the I2C bus and I found that after a few seconds the program/system would lock up, freeze the display and only be recovered by a hard reset.  My conditions:

- STM32F405 Feather Express
- SH1107 128x64 OLED FeatherWing (removed other devices from the bus)
- mu-editor on Ubuntu 18.04 LTS with Python 3.8
- CircuitPython 6.0.0-alpha.3-329-g19dbff67f-dirty

Is there a known issue with I2C?  Or this combination of stuff? Alpha CP? STM too new?

I also have seen the software not release the I2C properly (requires reset), which might be related to the above problem too.

-mark.

**Comment by mdroberts1243** [comment]
A small status update:
- I decided to build the quirk for the GrandCentral M4 Express board I bought.
- The SH1107 quirk built and ran fine with my driver
- I wasn't able to recreate the lockup/freeze that I saw on the STM32F405 Feather, even after hammering the I2C bus with continuous updates (moving "SH1107" around the screen with random increments)
- In separate testing I found I would see OS ErrNo 5 (I/O Error) on the STM32F405 when running the simpletest for the AS7341 as well.

So I'm concluding that there's an issue with the STM32F405 port and/or CircuitPython 6.0 alpha on the STM Feather.  I confirmed the STM I2C bus is at 400kHz and that is supported by the chip and the display controller.  I don't know what speed the GrandCentral M4 Express is running at, but it updates quite quickly.

**Comment by tannewt** [comment]
@mdroberts1243 I think those are the boards with the `_stage` module enabled. It calls set region to update itself here: https://github.com/adafruit/circuitpython/blob/main/shared-module/_stage/__init__.c#L46

I think you'll want to pull the parameter out of the display object like the other params.

**Comment by mdroberts1243** [comment]
@tannewt I made the change to _stage and rebuilt the Grand Central M4 Express just to make sure nothing was broken.  I had to resolve one conflict and merged it in. But now the checks are queued and seem to be stuck.  It is the checks that I want to see!

-mark.

**Comment by mdroberts1243** [comment]
@tannewt The checks failed.  Failing in a different way then before and the various board builds didn't run at all.  The only thing different was I was asked to resolve a conflict, which I did, accepting a bunch of changes to seeduino wio terminal apparently in order to add my quirk to the end of the display construct. In ports/atmel-samd/boards/seeeduino_wio_terminal/board.c 

Anyway, things seem to have regressed and I'm at a loss to interpret the failure.

File: shared-module/displayio/Display.c [hunk 7]
```diff
@@ -281,12 +290,18 @@ STATIC bool _refresh_area(displayio_display_obj_t* self, const displayio_area_t*
             .x2 = clipped.x2,
             .y2 = clipped.y1 + rows_per_buffer * (j + 1)
         };
-        if (remaining_rows < rows_per_buffer) {
+        if (self->SH1107_addressing) {
+            // one row only for SH1107 in vertical (column) mode
+            subrectangle.y1 = clipped.y1 + j;
+            subrectangle.y2 = clipped.y1 + (j + 1);
```
Comment (tannewt) [inline shared-module/displayio/Display.c:296]: Is this needed still? rows_per_buffer is 1 so it should be equivalent to the starting value.

**Review by tannewt — CHANGES_REQUESTED** [review]

**Comment by tannewt** [comment]
Note that the one build failure is a clone failure and can be ignored.

File: shared-module/displayio/Display.c (hunk 7 above)
Comment (mdroberts1243) [reply shared-module/displayio/Display.c:296]: Nope. Not required.  Tested on Grand Central M4 Express (slower than STM32F405... need to get that I2C solid!). Another commit coming...

**Review by tannewt — APPROVED** [review]
Thanks @mdroberts1243! This looks good.

**Comment by tannewt** [comment]
Driver repo is here: https://github.com/adafruit/Adafruit_CircuitPython_DisplayIO_SH1107
